### PR TITLE
`DefaultQubit2` preprocess the configs

### PR DESCRIPTION
**Pull request by albi3ro**
Follow on to PR #4007 . I wanted #4007  to focus that PR simply on the interface. This PR shows what behavior we will unlock with that change and what it will look like integrated.

### CODE REVIEW

File: tests/devices/experimental/test_default_qubit_2.py [hunk 1]
```diff
@@ -248,6 +282,8 @@ def test_tf_results_and_backprop(self):
 
 
 class TestExecutingBatches:
+    """Tests invovling executing multiple circuits at the same time."""
```
Comment (mudit2812) [inline tests/devices/experimental/test_default_qubit_2.py:285]:
```suggestion
    """Tests involving executing multiple circuits at the same time."""
```

**Review by mudit2812 — APPROVED** [review]
🙌

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -49,12 +49,21 @@
   [(#4079)](https://github.com/PennyLaneAI/pennylane/pull/4079)
   [(#4095)](https://github.com/PennyLaneAI/pennylane/pull/4095)
 
+* `qml.devices.ExecutionConfig` no longer has a `shots` property, as it is now on the `QuantumScript`.  It now has a `use_device_gradient` property. `ExecutionConfig.grad_on_execution = None` indicates a request for `"best"`, instead of a string.
+[(#4102)](https://github.com/PennyLaneAI/pennylane/pull/4102)
+
+* `DefaultQubit2.preprocess` now returns a new `ExecutionConfig` object will decisions for `gradient_method`,
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:55]:
```suggestion
* `DefaultQubit2.preprocess` now returns a new `ExecutionConfig` object with decisions for `gradient_method`,
```

**Review by mudit2812 — APPROVED** [review]

File: pennylane/devices/qubit/preprocess.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 """This module contains functions for preprocessing `QuantumTape` objects to ensure
 that they are supported for execution by a device."""
 # pylint: disable=protected-access
+import dataclasses
```
Comment (AmintorDusko) [inline pennylane/devices/qubit/preprocess.py:18]: I suggest in this case restricting the import to the functionality you are planning to use. This also helps to keep track of its usage. You can also rename `replace` if you feel this is a very generic name.
```suggestion
from dataclasses import replace
```

File: pennylane/devices/qubit/preprocess.py [hunk 4]
```diff
@@ -249,13 +273,9 @@ def preprocess(
             options for the execution.
 
     Returns:
-        Tuple[Sequence[.QuantumTape], callable]: Returns a tuple containing
-        the sequence of circuits to be executed, and a post-processing function
-        to be applied to the list of evaluated circuit results.
+        Tuple[QuantumTape], Callable, ExecutionConfig: QuantumTapes that the device can natively execute,
+        a postprocessing function to be called after execution, and a configuration with unset specifications filled in.
```
Comment (AmintorDusko) [inline pennylane/devices/qubit/preprocess.py:277]:
```suggestion
        a postprocessing function to be called after execution, and a configuration with originally unset specifications filled in.
```

File: pennylane/devices/qubit/preprocess.py [hunk 5]
```diff
@@ -249,13 +273,9 @@ def preprocess(
...
     """
-    if execution_config.shots is not None:
-        # Finite shot support will be added later
-        raise DeviceError("The Python Device does not support finite shots.")
```
Comment (AmintorDusko) [inline pennylane/devices/qubit/preprocess.py:258]: Do we have already support for finite shots on the new device API?

File: tests/devices/experimental/test_default_qubit_2.py [hunk 6]
```diff
@@ -381,6 +419,7 @@ class TestSumOfTermsDifferentiability:
 
     @staticmethod
     def f(scale, n_wires=10, offset=0.1, convert_to_hamiltonian=False):
+        """Execute a quantum script with a large hamiltonian."""
```
Comment (AmintorDusko) [inline tests/devices/experimental/test_default_qubit_2.py:422]:
```suggestion
        """Execute a quantum script with a large Hamiltonian."""
```

File: tests/devices/experimental/test_execution_config.py [hunk 7]
```diff
@@ -47,3 +48,16 @@ def test_invalid_gradient_keyword_arguments():
     """Tests that unknown gradient_keyword_arguments raise a ValueError."""
     with pytest.raises(ValueError, match="All gradient_keyword_arguments keys must be in"):
         _ = ExecutionConfig(gradient_keyword_arguments={"nonsense": 0})
+
+
+@pytest.mark.parametrize("option", (True, False, None))
+def test_valid_grad_on_execution(option):
+    """Test execution config allows the True, False and best"""
```
Comment (AmintorDusko) [inline tests/devices/experimental/test_execution_config.py:55]:
```suggestion
    """Test execution config allows True, False and None"""
```

File: tests/devices/qubit/test_preprocess.py [hunk 8]
```diff
@@ -55,12 +56,15 @@ class NoMatNoDecompOp(Operation):
...
     num_wires = 1
 
+    # pylint: disable=arguments-renamed, invalid-overridden-method
     @property
     def has_matrix(self):
         return False
 
 
 class TestPrivateHelpers:
+    """Test the privaate helpers for preprocessing."""
```
Comment (AmintorDusko) [inline tests/devices/qubit/test_preprocess.py:66]:
```suggestion
    """Test the private helpers for preprocessing."""
```

File: tests/devices/qubit/test_preprocess.py [hunk 9]
```diff
@@ -81,6 +85,7 @@ def test_accepted_operator(self, op, expected):
 
     @pytest.mark.parametrize("op", (qml.PauliX(0), qml.RX(1.2, wires=0), qml.QFT(wires=range(3))))
     def test_operator_decomposition_gen_accepted_operator(self, op):
+        """Test the operator decomposition gen function on an operator thats accepted."""
```
Comment (AmintorDusko) [inline tests/devices/qubit/test_preprocess.py:88]:
```suggestion
        """Test the _operator_decomposition_gen function on an operator that is accepted."""
```

File: tests/devices/qubit/test_preprocess.py [hunk 10]
```diff
@@ -191,6 +200,8 @@ def decomposition(self):
 
 
 class TestExpandFnTransformations:
+    """Tests for for the behavior of the `expand_fn` helper."""
```
Comment (AmintorDusko) [inline tests/devices/qubit/test_preprocess.py:203]:
```suggestion
    """Tests for the behavior of the `expand_fn` helper."""
```

**Review by AmintorDusko — APPROVED** [review]
A few suggestions, but looks good overall.
CIs are running successfully, and my suggestions are not expected to change that, so I will approve to not hold this PR any longer.

File: pennylane/devices/qubit/preprocess.py (hunk 5 above)
Comment (albi3ro) [reply pennylane/devices/qubit/preprocess.py:258]: shots are now on the `QuantumScript`, so they've been removed from the execution config.

A different PR in the near future will add preprocessing for finite shot situations.

File: pennylane/devices/qubit/preprocess.py [hunk 11]
```diff
@@ -234,10 +235,33 @@ def batch_fn(res):
...
+        updated_values["gradient_method"] = "backprop"
+    if config.use_device_gradient is None:
+        updated_values["use_device_gradient"] = config.gradient_method in {
+            "best",
+            "adjoint",
+            "backprop",
+        }
+    if config.grad_on_execution is None:
+        updated_values["grad_on_execution"] = config.gradient_method == "adjoint"
+    return dataclasses.replace(config, **updated_values)
```
Comment (albi3ro) [inline pennylane/devices/qubit/preprocess.py:258]:
```suggestion
    return replace(config, **updated_values)
```

File: pennylane/devices/qubit/preprocess.py (hunk 5 above)
Comment (AmintorDusko) [reply pennylane/devices/qubit/preprocess.py:258]: Thank you for that!

**Pull request merged**